src/librustc/util/ppaux.rs

```diff
@@ -1063,6 +1063,7 @@ define_print! {
                 TyParam(ref param_ty) => write!(f, "{}", param_ty),
                 TyAdt(def, substs) => cx.parameterized(f, substs, def.did, &[]),
                 TyDynamic(data, r) => {
+                    write!(f, "dyn ")?;
```

Comment:
Hmm, I think this may do the wrong thing in terms of precedence in some cases. For example, `&dyn Foo + 'a` needs some parens, right? I'm not 100% sure how this is handled already, since the same applies to `&Foo + 'a`... it kind of looks like we just get it wrong now?

Maybe we can concoct a test case to demonstrate this...